ADR-69 - Crowdsec IP filtering (#1761)
apache/james-project: src/adr/0069-crowdsec-IP-filtering.md

# 69. Crowdsec

Date: 2022-10-26

## Status

Accepted (lazy consensus).

Implemented. 

> ## Context

Reviewer: Why is fail2ban not enougth? (answers: cvery complex set up in a distributed architecture)

> Currently, there is no machanism to check harmful IP addresses implementing attacks like bruteforce attacks, dictionary attacks, etc.

Reviewer: Describe with one sentence each attack.

> To that end, James is integrated with [Crowdsec](https://www.crowdsec.net/). Crowdsec will check James logs, based on defined scenarios, Crowdsec can detect malevolent behaviors, and block them from accessing James at various levels (infrastructural, system, applicative). For basic, we implement Crowdsec to detect at system level.

Reviewer: This is a decision

A quick introduction about Crowdsec:

```
CrowdSec Security Engine is an open-source and lightweight software that allows you to detect peers with malevolent behaviors and block them from accessing your systems at various levels (infrastructural, system, applicative).

> To achieve this, the Security Engine reads logs from different sources (files, streams ...) to parse, normalize and enrich them before matching them to threats patterns called scenarios. 

Reviewer: Please add that crowdsec allow mutualizing data about security threats.

```

## Decision 

Set up a new maven project dedicated to Crowdsec extension. This allows to be embedded in a James server as a soft dependency
using the external-jar loading mechanism. With this way, the extension could be dropped in one's James installation, and not a runtime dependency.

To connect to Crowdsec, we use https protocol with reactor http client. 

## Consequences

> 

Reviewer: ?

Why empty

## Alternatives

> 

Reviewer: fail2ban?

# 69. Crowdsec

Date: 2022-10-26

## Status

Accepted (lazy consensus).

Implemented. 

## Context

Currently, there is no mechanism to check harmful IP addresses implementing attacks like bruteforce attacks, dictionary attacks, etc.

- Bruteforce attack: A hacker attempts to guess the user's login credentials.
- Dictionary attacks: A dictionary attack tries to find out the list of valid mail addresses.

Fail2ban can detect these attacks, however it is very complex to setup in a distributed system.

## Decision 

> To that end, James is integrated with [Crowdsec](https://www.crowdsec.net/). Crowdsec will check James logs, based on defined scenarios, Crowdsec can detect malevolent behaviors, and block them from accessing James at various levels (infrastructural, system, applicative). For basic, we implement Crowdsec to detect at system level.

Reviewer 2: I do not fully understand `various levels (infrastructural, system, applicative)`. Can you explain your idea regarding those 3 levels?

My impression is that James would leverage CrowdSec at the application layer (SMTP/IMAP).

Author: ![image](https://github.com/apache/james-project/assets/87489927/9521595a-1f85-496c-a028-e22d7389835e)
Remediation Components (Bouncer) 
`Remediation Components are software packages in charge of acting upon decisions provided by the Security Engine. Depending on where you would like to remediate the decision, you will need to install the appropriate remediation component.`

We can implement Remediation Components at various level.

Reviewer 2: How are the remediation levels mapped to `various levels (infrastructural, system, applicative)` then? Should unify the concepts, no?

Author: You mean that ` we implement Crowdsec to detect at system level.`  should be modified `at application level`, right? If so, I think you're right

A quick introduction about Crowdsec:

```
CrowdSec Security Engine is an open-source and lightweight software that allows you to detect peers with malevolent behaviors and block them from accessing your systems at various levels (infrastructural, system, applicative).

To achieve this, the Security Engine reads logs from different sources (files, streams ...) to parse, normalize and enrich them before matching them to threats patterns called scenarios. 

```

CrowdSec ships by default with scenarios (brute force, port scan, web scan, etc.) adapted for most contexts, but you can easily extend it by picking more of them from the [HUB](https://app.crowdsec.net/hub/collections). It is also easy to adapt an existing one or create one yourself.

Set up a new maven project dedicated to Crowdsec extension. This allows to be embedded in a James server as a soft dependency
using the external-jar loading mechanism. With this way, the extension could be dropped in one's James installation, and not a runtime dependency.

> James connects to Crowdsec via https protocol with reactor http client. 

Reviewer 2: What is the surface to be protected by CrowdSec? (IMAP, SMTP hook...)

Author: Yep, IMAP, SMTP and JMAP. They are all I can list down.

Reviewer 2: > Yep, IMAP, SMTP and JMAP. They are all I can list down.

I do not see those listed down yet.

Reviewer 2: > Crowdsec only cares for the content of logs, so instead of listing down the protocols, we should tell the abnormal behaviors that Crowdsec can detect.

It is CrowdSec responsibility to detect the attacks and ban decision. CrowdSec does not care much about application protocols or whatever except purely logs.

But which surfaces in James can leverage that ban decision? It is a different story. That should be noted in ADR.

BTW you did not precisely introduce the attack scenarios. ``james-bf-auth`, `dictionary-attack`` are just the names in your POC, those can change in any time. Should not force reader to guess `bf` is bruteforce, no?

Let's list the attacks FQDN and explain them a bit IMO.

Author: Thank you for your explanation

# 69. Crowdsec

Date: 2022-10-26

## Status

Accepted (lazy consensus).

Implemented. 

## Context

Currently, there is no mechanism to check harmful IP addresses implementing attacks like bruteforce attacks, dictionary attacks, etc.

- Bruteforce attack: A hacker attempts to guess the user's login credentials.
- Dictionary attacks: A dictionary attack tries to find out the list of valid mail addresses.

## Decision 

> To that end, James is integrated with [Crowdsec](https://www.crowdsec.net/). Crowdsec will check James logs, based on defined scenarios, Crowdsec can detect malevolent behaviors, and block them from accessing James at various levels (infrastructural, system, applicative). For basic, we implement Crowdsec to detect at application level.

Reviewer: ```suggestion
To that end, we will integrate James with [Crowdsec](https://www.crowdsec.net/). Crowdsec will check James logs, based on defined scenarios, Crowdsec can detect malevolent behaviors, and block them from accessing James at various levels (infrastructure - firewall, system, applicative). To ease end user configuration, and propose a bundled experience indepandant from end user infrastructure, we implement Crowdsec remediations at application level. Willing users can very well set up remediations based on their underlying infrastructure.
```

using the external-jar loading mechanism. With this way, the extension could be dropped in one's James installation, and not a runtime dependency.

### How does Crowdsec apply?
- Many attacks on James via IMAP, SMTP, JMAP can be detected by Crowdsec
- Crowdsec analyzes logs from James, then based on their patterns in parsers, it can detect each behavior in James. 
- Scenarios are YAML files that allow to detect a specific behavior, usually an attack. With James, we detect bruteforce login attack and Dictionary attack.![img.png](img/adr-69-scenario-diagram.png)
- Each line log is an event in Crowdsec. If an event triggers the condition in Crowdsec, Crowdsec will send an alert to the Local API, and make a decision (ban)
- James connects to Crowdsec via https protocol with reactor http client to query the decision from Crowdsec

## Consequences

- James will be more secure by blocking threats at various levels (applicative, system, infrastructural).
- We can document and share Crowdsec scenario thus allowing users to benefit from each other decisions. Also we can get dangerous IP detected by community via using Central API 

> 

Reviewer: ```suggestion
- Crowdsec natively supports distributed deployments (reference: ) by setting up an agent per node gathering logs of local containers and forwarding them to the crowdsec API that decides if remediations should be applied.
This architecture is not horizontally scalable in itself but should already be able to analyse over 1000 relevant logs per seconds (metric obtained by discussing with Crowdsec team), which should allow leveraging very large scale James deployments. 

```

## Alternatives

> - Fail2ban can detect these attacks, however it is complex to setup in a distributed system.

Reviewer: ```suggestion
- Fail2ban can detect these attacks, however it is complex to setup in a distributed system and very related to the underlying architecture thus hard to document within the community.
```

- Fail2ban do not permit sharing with peers the underlying blocked data.

> 

Reviewer: We could include a section for complementary work...

Ideas:

Additional triggers:
 - User exceed their sending rate. This could be done for instance via a mailet that emits a log incorporating

Additional remediation:
 - Matchers that queries crowdsec for a given sender ? This could be used for remediation, eg if a user sends too much mails...
 - Have a MailFromHook querying crowdsec for the sender? 
 - Other "softer" remediations in SMTP might include Grey Listing (once implemented).

Author: Maybe in another commit?

Reviewer: No here?

## Context

Currently, there is no mechanism to check harmful IP addresses implementing attacks like bruteforce attacks, dictionary attacks, etc.

- Bruteforce attack: A hacker attempts to guess the user's login credentials.
- Dictionary attacks: A dictionary attack tries to find out the list of valid mail addresses.

> Fail2ban can detect these attacks, however it is very complex to setup in a distributed system.

Reviewer: ```suggestion
```

CrowdSec Security Engine is an open-source and lightweight software that allows you to detect peers with malevolent behaviors and block them from accessing your systems at various levels (infrastructural, system, applicative).

To achieve this, the Security Engine reads logs from different sources (files, streams ...) to parse, normalize and enrich them before matching them to threats patterns called scenarios. 

```

CrowdSec ships by default with scenarios (brute force, port scan, web scan, etc.) adapted for most contexts, but you can easily extend it by picking more of them from the [HUB](https://app.crowdsec.net/hub/collections). It is also easy to adapt an existing one or create one yourself.

Set up a new maven project dedicated to Crowdsec extension. This allows to be embedded in a James server as a soft dependency
using the external-jar loading mechanism. With this way, the extension could be dropped in one's James installation, and not a runtime dependency.

James connects to Crowdsec via https protocol with reactor http client. 

## Consequences

> James will be more secure by blocking threats at various levels (applicative, system, infrastructural). 

Reviewer: - We can document and share Crowdsec scenari thus allowing users to benefit from each other decisions.

## Alternatives

> - Fail2ban

Reviewer: ```suggestion

Fail2ban can detect these attacks, however it is complex to setup in a distributed system. 
Fail2ban do not permit sharing with peers the underlying blocked data.
```
